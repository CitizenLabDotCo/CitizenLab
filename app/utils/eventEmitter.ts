import * as Rx from 'rxjs/Rx';

interface IEventEmitterEvent<T> {
  eventSource: string;
  eventName: string;
  eventValue: T;
}

class EventEmitter {
  private subject: Rx.Subject<IEventEmitterEvent<any>>;
  private stream: { [key: string]: Rx.Observable<IEventEmitterEvent<any>> };

  constructor() {
    this.subject = new Rx.Subject();
    this.stream = {};
  }

  emit<T>(eventSource: string, eventName: string, eventValue: T) {
    this.subject.next({ eventSource, eventName, eventValue });
  }

  observeEventFromSource<T>(eventSource: string, eventName: string): Rx.Observable<IEventEmitterEvent<T>> {
    const streamName = `${eventSource}-${eventName}`;

    if (!this.stream[streamName]) {
      this.stream[streamName] = this.subject.filter(data => data.eventSource === eventSource && data.eventName === eventName).share();
    }

    return this.stream[streamName];
  }

  observeEvent<T>(eventName: string): Rx.Observable<IEventEmitterEvent<T>> {
    const streamName = eventName;

    if (!this.stream[streamName]) {
      this.stream[streamName] = this.subject.filter(data => data.eventName === eventName).share();
    }

    return this.stream[streamName];
  }
}

const eventEmitter = new EventEmitter();
export default eventEmitter;
