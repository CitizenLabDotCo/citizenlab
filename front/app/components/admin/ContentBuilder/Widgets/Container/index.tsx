import React from 'react';

import { Box } from '@citizenlab/cl2-component-library';
import { UserComponent } from '@craftjs/core';

import Events from 'containers/Admin/pagesAndMenu/containers/ContentBuilder/components/CraftComponents/Events';
import Projects from 'containers/Admin/pagesAndMenu/containers/ContentBuilder/components/CraftComponents/Projects';
import Proposals from 'containers/Admin/pagesAndMenu/containers/ContentBuilder/components/CraftComponents/Proposals';

const Container: UserComponent = ({ children }) => {
  return (
    <Box className="e2e-single-column" minHeight="40px" w="100%">
      {children}
    </Box>
  );
};

Container.craft = {
  rules: {
    canMoveIn: (incomingNodes) => {
      return !incomingNodes.some(
        (node) =>
          node.data.type === Events ||
          node.data.type === Proposals ||
          node.data.type === Projects
      );
    },
  },
};

export default Container;
