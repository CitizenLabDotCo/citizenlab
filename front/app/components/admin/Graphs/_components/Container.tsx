import React, { useMemo } from 'react';
import { useResizeDetector } from 'react-resize-detector';

// components
import { ResponsiveContainer } from 'recharts';
import FakeLegend from './Legend/FakeLegend';
import { Box } from '@citizenlab/cl2-component-library';

// utils
import { debounce, isEqual } from 'lodash-es';

// typings
import { Percentage } from 'typings';
import {
  GraphDimensions,
  LegendDimensions,
} from '../_components/Legend/typings';
import { Legend } from '../typings';

interface Props {
  width?: number | Percentage;
  height?: number | Percentage;
  legend?: Legend;
  graphDimensions: GraphDimensions | undefined;
  legendDimensions: LegendDimensions | undefined;
  defaultLegendOffset: number;
  onUpdateGraphDimensions?: (graphDimensions: GraphDimensions) => void;
  onUpdateLegendDimensions?: (legendDimensions: LegendDimensions) => void;
  children: React.ReactElement;
}

const warn = console.warn.bind(console);
function wrapWarn(...args) {
  if (
    typeof args[0] === 'string' &&
    (args[0].includes('ResponsiveContainer') ||
      args[0].includes('minWidth(undefined) or minHeight(undefined) '))
  ) {
    return;
  }
  warn(...args);
}

console.warn = wrapWarn;

const Container = ({
  width,
  height,
  legend,
  graphDimensions,
  legendDimensions,
  defaultLegendOffset,
  onUpdateGraphDimensions,
  onUpdateLegendDimensions,
  children,
}: Props) => {
  const handleResize = useMemo(
    () =>
      debounce((width: number, height: number) => {
        const newGraphDimensions = { width, height };

        if (!isEqual(graphDimensions, newGraphDimensions)) {
          if (onUpdateGraphDimensions) {
            onUpdateGraphDimensions(newGraphDimensions);
          }
        }
      }, 50),
    [graphDimensions, onUpdateGraphDimensions]
  );

  const { ref: resizeRef } = useResizeDetector({ onResize: handleResize });

  const handleRef = (ref: React.RefObject<HTMLDivElement>) => {
    if (ref === null) return;
    const node = ref.current;
    if (node === null) return;

    resizeRef.current = node;
  };

  const rightLegend = legend?.position?.includes('right');
  const maintainGraphSize = !!legend?.maintainGraphSize;

  const parsedWidth =
    rightLegend && typeof width === 'number' && maintainGraphSize
      ? width +
        (legendDimensions?.width ?? 0) +
        (legend.marginLeft ?? defaultLegendOffset)
      : width;

  const parsedHeight =
    !rightLegend && typeof height === 'number' && maintainGraphSize
      ? height +
        (legendDimensions?.height ?? 0) +
        (legend.marginTop ?? defaultLegendOffset)
      : height;

  return (
    <Box
      display="flex"
      flexDirection={rightLegend ? 'row' : 'column'}
      justifyContent="center"
      width="100%"
      height="100%"
    >
      <ResponsiveContainer
        width={parsedWidth}
        height={parsedHeight}
        ref={handleRef}
      >
        {children}
      </ResponsiveContainer>

      {legend && onUpdateLegendDimensions && (
        <FakeLegend
          width={width}
          items={legend.items}
          position={legend.position}
          onCalculateDimensions={onUpdateLegendDimensions}
        />
      )}
    </Box>
  );
};

export default Container;
