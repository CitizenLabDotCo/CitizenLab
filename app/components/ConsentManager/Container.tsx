import React, { PureComponent } from 'react';

import Banner from './Banner';
import Preferences from './Preferences';
import Modal from 'components/UI/Modal';
import Button from 'components/UI/Button';

import { ADVERTISING_CATEGORIES, FUNCTIONAL_CATEGORIES } from './categories';

import { IDestination } from 'utils/analytics';

interface Props {
  setPreferences: Function;
  resetPreferences: Function;
  saveConsent: Function;
  destinations: IDestination[];
  newDestinations: IDestination[];
  preferences: any;
  isConsentRequired: boolean;
  implyConsentOnInteraction: boolean;
}

interface State {
  isDialogOpen: boolean;
  isCancelling: boolean;
}

export default class Container extends PureComponent<Props, State> {
  constructor(props) {
    super(props);
    this.state = {
      isDialogOpen: false,
      isCancelling: false
    };
  }

  render() {
    console.log(this.state);
    const {
      destinations,
      newDestinations,
      preferences,
      isConsentRequired,
    } = this.props;
    const { isDialogOpen, isCancelling } = this.state;
    const marketingDestinations = [] as IDestination[];
    const advertisingDestinations = [] as IDestination[];
    const functionalDestinations = [] as IDestination[];

    for (const destination of destinations) {
      if (ADVERTISING_CATEGORIES.find(c => c === destination.category)) {
        advertisingDestinations.push(destination);
      } else if (FUNCTIONAL_CATEGORIES.find(c => c === destination.category)) {
        functionalDestinations.push(destination);
      } else {
        // Fallback to marketing
        marketingDestinations.push(destination);
      }
    }

    // TODO: add state for banner so it doesn't disappear on implicit consent (which is annoying UX)
    return (
      <>
        {isConsentRequired &&
          newDestinations.length > 0 && (
            <Banner
              onAccept={this.handleBannerAccept}
              onChangePreferences={this.openDialog}
            />
          )}
        <Modal
          opened={isDialogOpen || isCancelling}
          close={this.closeDialog}
          header={<h2>Your privacy preferences</h2>}
          footer={this.renderPrefFooter(preferences.marketingAndAnalytics, preferences.advertising, preferences.functional, isDialogOpen, isCancelling)}
        >
          {isDialogOpen &&
            <Preferences
              onCancel={this.handleCancel}
              onSave={this.handleSave}
              onChange={this.handleCategoryChange}
              marketingDestinations={marketingDestinations}
              advertisingDestinations={advertisingDestinations}
              functionalDestinations={functionalDestinations}
              marketingAndAnalytics={preferences.marketingAndAnalytics}
              advertising={preferences.advertising}
              functional={preferences.functional}
            />
          }
          {isCancelling &&
            <div>
              If you cancel, your preferences won't be saved
            </div>
          }
        </Modal>
      </>
    );
  }
  renderPrefFooter = (marketingAndAnalytics, advertising, functional, isDialogOpen, isCancelling) => {
    if (isDialogOpen) {
      return (
        <div>
          <Button onClick={this.handleCancel}>
            Cancel
      </Button>
          <Button onClick={this.handleSubmit(marketingAndAnalytics, advertising, functional)}>Save</Button>
        </div>
      );
    } else if (isCancelling) {
      return (
        <div>
          <Button onClick={this.handleCancelBack}>
            Go back
      </Button>
          <Button onClick={this.handleCancelConfirm}>Cancel anyway</Button>
        </div>
      );
    }
    return;
  }
  handleSubmit = (marketingAndAnalytics, advertising, functional) => e => {

    e.preventDefault();

    // Safe guard against browsers that don't prevent the
    // submission of invalid forms (Safari < 10.1)
    if (
      marketingAndAnalytics === null ||
      advertising === null ||
      functional === null
    ) {
      return;
    }

    this.handleSave();
  }

  openDialog = () => {
    console.log('open sesame');
    this.setState({
      isDialogOpen: true
    });
  }

  closeDialog = () => {
    this.setState({
      isDialogOpen: false
    });
  }

  handleBannerAccept = () => {
    const { saveConsent } = this.props;

    saveConsent();
  }

  handleCategoryChange = (category, value) => {
    const { setPreferences } = this.props;

    setPreferences({
      [category]: value
    });
  }

  handleSave = () => {
    const { saveConsent } = this.props;

    this.setState({
      isDialogOpen: false
    });
    saveConsent();
  }

  handleCancel = () => {
    const { resetPreferences, newDestinations } = this.props;

    this.setState({
      isDialogOpen: false
    });

    // Only show the cancel confirmation if there's unconsented destinations
    if (newDestinations.length > 0) {
      this.setState({
        isCancelling: true
      });
    } else {
      resetPreferences();
    }
  }

  handleCancelBack = () => {
    this.setState({
      isDialogOpen: true,
      isCancelling: false
    });
  }

  handleCancelConfirm = () => {
    const { resetPreferences } = this.props;

    this.setState({
      isCancelling: false
    });
    resetPreferences();
  }
}
