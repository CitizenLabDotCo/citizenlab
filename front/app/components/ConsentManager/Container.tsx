import React, { PureComponent, FormEvent } from 'react';
import { Subscription } from 'rxjs';

// Events
import eventEmitter from 'utils/eventEmitter';

// Components
import Banner from './Banner';
import PreferencesDialog, { ContentContainer } from './PreferencesDialog';
import Footer from './Footer';
import LoadableModal from 'components/Loadable/Modal';

import { FormattedMessage } from 'utils/cl-intl';
import messages from './messages';

import { CategorizedDestinations, IPreferences } from './';

import styled from 'styled-components';

export const ButtonContainer = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;
`;

interface Props {
  setPreferences: (arg: Record<string, unknown>) => void;
  resetPreferences: () => void;
  accept: () => void;
  saveConsent: () => void;
  isConsentRequired: boolean;
  preferences: IPreferences;
  categorizedDestinations: CategorizedDestinations;
  onToggleModal: (opened: boolean) => void;
}

interface State {
  isDialogOpen: boolean;
  isCancelling: boolean;
}

export default class Container extends PureComponent<Props, State> {
  subscriptions: Subscription[] = [];

  constructor(props) {
    super(props);
    this.state = {
      isDialogOpen: false,
      isCancelling: false,
    };
  }

  componentDidMount() {
    this.subscriptions = [
      eventEmitter
        .observeEvent('openConsentManager')
        .subscribe(this.openDialog),
    ];
  }

  componentWillUnmount() {
    this.subscriptions.forEach((subscription) => subscription.unsubscribe());
  }

  openDialog = () => {
    this.props.onToggleModal(false);
    this.setState({
      isDialogOpen: true,
    });
  };

  closeDialog = () => {
    this.props.onToggleModal(true);
    this.setState({
      isDialogOpen: false,
    });
  };

  handleCategoryChange = (category: string, value: boolean) => {
    const { setPreferences } = this.props;

    setPreferences({
      [category]: value,
    });
  };

  validate = () => {
    let res = true;
    const { preferences, categorizedDestinations } = this.props;
    for (const category of Object.keys(categorizedDestinations)) {
      if (categorizedDestinations[category].length > 0) {
        res = res && !(preferences[category] === undefined);
      }
    }
    return res;
  };

  handleSave = (e: FormEvent<any>) => {
    e.preventDefault();

    if (!this.validate()) {
      return;
    }

    const { saveConsent } = this.props;

    this.setState({
      isDialogOpen: false,
    });
    saveConsent();
  };

  handleCancel = () => {
    const { resetPreferences, isConsentRequired, preferences } = this.props;

    const isEmpty = Object.values(preferences).every((e) => e === undefined);

    // Only show the cancel confirmation if there's unconsented destinations...
    // or if the user made a choice and we want to confirm aborting it
    if (isConsentRequired && !isEmpty) {
      this.setState({ isCancelling: true });
    } else {
      this.setState({ isDialogOpen: false });
      resetPreferences();
    }
  };

  handleCancelBack = () => {
    this.setState({ isCancelling: false });
  };

  handleCancelConfirm = () => {
    const { resetPreferences } = this.props;

    this.setState({
      isCancelling: false,
      isDialogOpen: false,
    });
    resetPreferences();
  };

  render() {
    const {
      preferences,
      isConsentRequired,
      categorizedDestinations,
      accept,
    } = this.props;
    const { isDialogOpen, isCancelling } = this.state;
    const noDestinations = Object.values(categorizedDestinations).every(
      (array) => array.length === 0
    );
    const mode = noDestinations
      ? 'noDestinations'
      : !isCancelling
      ? 'preferenceForm'
      : 'cancelling';

    return (
      <>
        <LoadableModal
          opened={isDialogOpen}
          close={this.closeDialog}
          header={<FormattedMessage {...messages.title} />}
          footer={
            <Footer
              validate={this.validate}
              mode={mode}
              handleCancelBack={this.handleCancelBack}
              handleCancelConfirm={this.handleCancelConfirm}
              handleCancel={this.handleCancel}
              handleSave={this.handleSave}
            />
          }
        >
          {!isCancelling ? (
            <PreferencesDialog
              onChange={this.handleCategoryChange}
              categoryDestinations={categorizedDestinations}
              preferences={preferences}
            />
          ) : (
            <ContentContainer role="dialog" aria-modal>
              <FormattedMessage {...messages.confirmation} tagName="h1" />
            </ContentContainer>
          )}
        </LoadableModal>

        {isConsentRequired && (
          <Banner onAccept={accept} onChangePreferences={this.openDialog} />
        )}
      </>
    );
  }
}
