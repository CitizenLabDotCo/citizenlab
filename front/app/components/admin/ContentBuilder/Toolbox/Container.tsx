import React from 'react';

// components
import { Box } from '@citizenlab/cl2-component-library';

// styling
import { colors } from 'utils/styleUtils';

interface Props {
  children: React.ReactNode;
}

const Container = ({ children }: Props) => (
  <Box
    position="fixed"
    zIndex="99999"
    flex="0 0 auto"
    h="100%"
    w="210px"
    display="flex"
    flexDirection="column"
    alignItems="center"
    bgColor="#ffffff"
    overflowY="auto"
    borderRight={`1px solid ${colors.grey500}`}
  >
    <Box w="100%" display="inline">
      {children}
    </Box>
  </Box>
);

export default Container;
