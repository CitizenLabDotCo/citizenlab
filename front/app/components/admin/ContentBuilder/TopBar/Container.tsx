import React from 'react';

// components
import { Box } from '@citizenlab/cl2-component-library';

// stying
import { colors, stylingConsts } from 'utils/styleUtils';

interface Props {
  children: React.ReactNode;
}

const Container = ({ children }: Props) => (
  <Box
    position="fixed"
    zIndex="3"
    alignItems="center"
    w="100%"
    h={`${stylingConsts.menuHeight}px`}
    display="flex"
    background={`${colors.white}`}
    borderBottom={`1px solid ${colors.grey500}`}
  >
    {children}
  </Box>
);

export default Container;
