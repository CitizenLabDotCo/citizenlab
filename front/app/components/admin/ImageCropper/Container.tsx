import React from 'react';
import { Text, Box } from '@citizenlab/cl2-component-library';
import Warning from 'components/UI/Warning';
import { FormattedMessage, useIntl } from 'utils/cl-intl';
import ImageCropper, { ImageCropperProps } from '.';
import messages from './messages';

const Container = ({ aspect, ...otherProps }: ImageCropperProps) => {
  const { formatMessage } = useIntl();
  return (
    <Box>
      <ImageCropper aspect={aspect} {...otherProps} />
      <Warning>
        <Text>
          <FormattedMessage
            {...messages.info}
            values={{
              link: (
                <a
                  href={formatMessage(messages.imageSupportPageURL)}
                  target="_blank"
                  rel="noreferrer"
                >
                  <FormattedMessage {...messages.infoLinkText} />
                </a>
              ),
              aspect: aspect,
            }}
          />
        </Text>
      </Warning>
    </Box>
  );
};

export default Container;
