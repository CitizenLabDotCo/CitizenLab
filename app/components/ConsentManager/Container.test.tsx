// libraries
import React from 'react';
import { shallow } from 'enzyme';
import eventEmitter from 'utils/eventEmitter';

import 'jest-styled-components';

// component to test
import { Container } from './Container';
import { initialPreferences } from './';

// mock utilities
jest.mock('./Banner', () => 'Banner');
jest.mock('./PreferencesDialog', () => 'PreferencesDialog');
jest.mock('./Footer', () => 'Footer');
jest.mock('components/Loadable/Modal', () => 'LoadableModal');

jest.mock('utils/cl-intl');
const Intl = require('utils/cl-intl/__mocks__/');
const { intl } = Intl;

const categorizedDestinations = {
  advertising:
    [{
      name: 'Google Tag Manager',
      description:
        'Google Tag Manager is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.',
      category: 'Tag Managers',
      website: 'http://google.com/tagManager',
      id: 'Google Tag Manager'
    },
    {
      name: 'AdvertisingTool',
      description: 'Advertising BS',
      category: 'Advertising',
      website: 'http://random.com/advertising',
      id: 'AdvertisingTool'
    }],
  functional:
    [{
      name: 'FunctionalTool',
      description: 'Actually might be handy',
      category: 'Security & Fraud',
      website: 'http://random.com/securitycookie',
      id: 'FunctionalTool'
    }],
  analytics:
    [{
      name: 'MarketingTool',
      description:
        'MarketingTool is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.',
      category: 'Analytics',
      website: 'http://random.com/marketing',
      id: 'MarketingTool'
    }]
};
let setPreferences: jest.Mock;
let resetPreferences: jest.Mock;
let saveConsent: jest.Mock;

describe('<Container />', () => {
  beforeEach(() => {
    setPreferences = jest.fn();
    resetPreferences = jest.fn();
    saveConsent = jest.fn();
  });

  it('lets the user accept all from the banner', () => {
    const wrapper = shallow(
      <Container
        intl={intl}
        setPreferences={setPreferences}
        resetPreferences={resetPreferences}
        saveConsent={saveConsent}
        isConsentRequired={true}
        preferences={initialPreferences}
        categorizedDestinations={categorizedDestinations}
      />
    );

    wrapper.find('Banner').props().onAccept();

    // accept all only calls saveConsent with preferences unchanged, so still equal to initialPreferences.
    // it's tested on the ConsentManager that this will have the expected behaviour.
    expect(setPreferences).not.toHaveBeenCalled();
    expect(saveConsent).toHaveBeenCalled();
    expect(resetPreferences).not.toHaveBeenCalled();
  });
  it('renders correclty when no destinations are allowed by the tenant', () => {
    const wrapper = shallow(
      <Container
        intl={intl}
        setPreferences={setPreferences}
        resetPreferences={resetPreferences}
        saveConsent={saveConsent}
        isConsentRequired={false}
        preferences={initialPreferences}
        categorizedDestinations={{
          analytics: [],
          functional: [],
          advertising: []
        }}
      />
    );

    // wih no destinations allowed, isConsentRequired will be false so no banner
    // but the modal is still accessible trough the cookie policy
    eventEmitter.emit('openConsentManager');
    expect(wrapper).toMatchSnapshot();
  });

  it('handle changes as expected', () => {
    const wrapper = shallow(
      <Container
        intl={intl}
        setPreferences={setPreferences}
        resetPreferences={resetPreferences}
        saveConsent={saveConsent}
        isConsentRequired={true}
        preferences={initialPreferences}
        categorizedDestinations={categorizedDestinations}
      />
    );

    wrapper.find('PreferencesDialog').props().onChange('advertising', false);

    // accept all only calls saveConsent with preferences unchanged, so still equal to initialPreferences.
    // it's tested on the ConsentManager that this will have the expected behaviour.
    expect(setPreferences).toHaveBeenCalledWith({ advertising: false });
    expect(saveConsent).not.toHaveBeenCalled();
    expect(resetPreferences).not.toHaveBeenCalled();
  });

  describe('shows the banner is and only if consent is required', () => {
    it('consent is required, it shows the banner', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={initialPreferences}
          categorizedDestinations={categorizedDestinations}
        />
      );
      expect(wrapper.find('Banner').exists()).toBeTruthy();
    });
    it('consent is\'t required, it doesn\'t show the banner', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={false}
          preferences={initialPreferences}
          categorizedDestinations={categorizedDestinations}
        />
      );
      expect(wrapper.find('Banner').exists()).toBeFalsy();
    });
  });

  describe('handles opening and closing the modal', () => {
    it('modal is initially closed when consent is required', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={initialPreferences}
          categorizedDestinations={categorizedDestinations}
        />
      );
      expect(wrapper.find('LoadableModal').props().opened).toBe(false);
    });
    it('modal is initially closed when consent is\'t required', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={false}
          preferences={initialPreferences}
          categorizedDestinations={categorizedDestinations}
        />
      );
      expect(wrapper.find('LoadableModal').props().opened).toBe(false);
    });
    it('passes down to Banner a handler that opens the modal', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={initialPreferences}
          categorizedDestinations={categorizedDestinations}
        />
      );
      expect(wrapper.find('LoadableModal').props().opened).toBe(false);
      wrapper.find('Banner').props().onChangePreferences();
      expect(wrapper.find('LoadableModal').props().opened).toBe(true);
    });

    it('passes down to Modal a handler that closes the modal', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={initialPreferences}
          categorizedDestinations={categorizedDestinations}
        />
      );
      wrapper.instance().setState({ isDialogOpen: true });
      expect(wrapper.find('LoadableModal').props().opened).toBe(true);
      wrapper.find('LoadableModal').props().close();
      expect(wrapper.find('LoadableModal').props().opened).toBe(false);
    });

    it('reacts to openConsentManager events by opening the modal', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={initialPreferences}
          categorizedDestinations={categorizedDestinations}
        />
      );
      expect(wrapper.find('LoadableModal').props().opened).toBe(false);
      eventEmitter.emit('openConsentManager');
      expect(wrapper.find('LoadableModal').props().opened).toBe(true);
    });
  });

  describe('handles preferences form validation', () => {
    it('is invalid when when some field is empty', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={{
            analytics: true,
            functional: null,
            advertising: null
          }}
          categorizedDestinations={categorizedDestinations}
        />
      );
      expect((wrapper.instance() as any).validate()).toBeFalsy();
    });
    it('is valid when when no field is empty', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={{
            analytics: true,
            functional: false,
            advertising: false
          }}
          categorizedDestinations={categorizedDestinations}
        />
      );
      expect((wrapper.instance() as any).validate()).toBeTruthy();
    });
  });

  describe('handles cancelling', () => {
    it('cancelling with no preference set closes the modal and lets the banner open', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={initialPreferences}
          categorizedDestinations={categorizedDestinations}
        />
      );
      wrapper.instance().setState({ isDialogOpen: true });

      expect(wrapper.find('Banner').exists()).toBeTruthy();
      expect(wrapper.find('LoadableModal').props().opened).toBe(true);
      const { footer } = wrapper.find('LoadableModal').props();
      shallow(footer).props().handleCancel();
      expect(wrapper.find('LoadableModal').props().opened).toBe(false);
      expect(wrapper.find('Banner').exists()).toBeTruthy();
      expect(setPreferences).not.toHaveBeenCalled();
      expect(saveConsent).not.toHaveBeenCalled();
      expect(resetPreferences).toHaveBeenCalled();
    });

    it('cancelling with some preference set prompts confirmation and lets the banner open', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={{
            analytics: true,
            functional: null,
            advertising: null
          }}
          categorizedDestinations={categorizedDestinations}
        />
      );
      wrapper.instance().setState({ isDialogOpen: true });

      expect(wrapper.find('Banner').exists()).toBeTruthy();
      expect(wrapper.find('LoadableModal').props().opened).toBe(true);
      const { footer } = wrapper.find('LoadableModal').props();
      shallow(footer).props().handleCancel();
      expect(wrapper.find('LoadableModal').props().opened).toBe(true);
      expect(wrapper.state().isCancelling).toBe(true);
      expect(wrapper.find('Banner').exists()).toBeTruthy();
      expect(setPreferences).not.toHaveBeenCalled();
      expect(saveConsent).not.toHaveBeenCalled();
      expect(resetPreferences).not.toHaveBeenCalled();
    });

    it('while cancelling, confirming cancellation closes the modal and keeps the banner', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={{
            analytics: true,
            functional: null,
            advertising: null
          }}
          categorizedDestinations={categorizedDestinations}
        />
      );
      wrapper.instance().setState({ isDialogOpen: true, isCancelling: true });

      expect(wrapper.find('Banner').exists()).toBeTruthy();
      const { footer } = wrapper.find('LoadableModal').props();
      shallow(footer).props().handleCancelConfirm();
      expect(wrapper.find('LoadableModal').props().opened).toBe(false);
      expect(wrapper.find('Banner').exists()).toBeTruthy();
      expect(setPreferences).not.toHaveBeenCalled();
      expect(saveConsent).not.toHaveBeenCalled();
      expect(resetPreferences).toHaveBeenCalled();
    });

    it('while cancelling, cancelling cancellation is still possible', () => {
      const wrapper = shallow(
        <Container
          intl={intl}
          setPreferences={setPreferences}
          resetPreferences={resetPreferences}
          saveConsent={saveConsent}
          isConsentRequired={true}
          preferences={{
            analytics: true,
            functional: null,
            advertising: null
          }}
          categorizedDestinations={categorizedDestinations}
        />
      );
      wrapper.instance().setState({ isDialogOpen: true, isCancelling: true });

      expect(wrapper.find('Banner').exists()).toBeTruthy();
      const { footer } = wrapper.find('LoadableModal').props();
      shallow(footer).props().handleCancelBack();
      expect(wrapper.find('LoadableModal').props().opened).toBe(true);
      expect(wrapper.find('Banner').exists()).toBeTruthy();
      expect(setPreferences).not.toHaveBeenCalled();
      expect(saveConsent).not.toHaveBeenCalled();
      expect(resetPreferences).not.toHaveBeenCalled();
    });
  });
});
