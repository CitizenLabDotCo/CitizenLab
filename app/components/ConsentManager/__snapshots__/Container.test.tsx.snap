// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> renders correclty when no destinations are allowed by the tenant 1`] = `
<Fragment>
  <LoadableModal
    close={[Function]}
    footer={
      <Footer
        handleCancel={[Function]}
        handleCancelBack={[Function]}
        handleCancelConfirm={[Function]}
        handleSave={[Function]}
        mode="noDestinations"
        validate={[Function]}
      />
    }
    header={
      <FormattedMessage
        defaultMessage="Your cookie preferences"
        id="app.components.ConsentManager.Modal.PreferencesDialog.title"
      />
    }
    opened={true}
  >
    <Component
      aria-modal={true}
      role="dialog"
    >
      <FormattedMessage
        defaultMessage="No trackers are enabled on your platform"
        id="app.components.ConsentManager.Banner.noDestinations"
        tagName="h1"
      />
    </Component>
  </LoadableModal>
</Fragment>
`;
